### PR TITLE
Gateway streaming for Cohere

### PR DESCRIPTION
**Pull request by gabrielfu**
<details><summary>&#x1F6E0 DevTools &#x1F6E0</summary>
<p>

[Open in GitHub Codespaces](https://codespaces.new/gabrielfu/mlflow/pull/10856?quickstart=1)

#### Install mlflow from this PR

```
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/10856/merge
```

#### Checkout with GitHub CLI

```
gh pr checkout 10856
```

</p>
</details>

### Related Issues/PRs

 #xxx

### What changes are proposed in this pull request?

Implement streaming completions for Cohere

### How is this PR tested?

- [X] Existing unit/integration tests
- [ ] New unit/integration tests
- [x] Manual tests

Manual test:

```yaml
# config.yaml
routes:
  - name: cohere
    route_type: llm/v1/completions
    model:
      provider: cohere
      name: command
      config:
        cohere_api_key: <key>
```

```shell
mlflow deployments start-server --config-path config.yaml --workers 1
```

```shell
curl -X POST http://127.0.0.1:5000/endpoints/cohere/invocations \
  -H "Content-Type: application/json" \
  -d '{"prompt": "Hello", "stream": true}'
```

### Does this PR require documentation update?

- [ ] No. You can skip the rest of this section.
- [X] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [x] Instructions

#### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [X] Yes. Give a description of this change to be included in the release notes for MLflow users.

Add support for streaming completions route for Cohere

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [X] `area/deployments`: MLflow Deployments client APIs, server, and third-party Deployments integrations
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [X] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: mlflow/gateway/providers/cohere.py [hunk 1]
```diff
@@ -45,6 +46,63 @@ def model_to_completions(cls, resp, config):
...
+        #               "text":" Hi there, how can I assist you today? ",
+        #               "finish_reason":"COMPLETE"
+        #           }
+        #       ],
+        #       "prompt":"Hello"
+        #   }}
+        # ```
+        response = resp.get("response")
+        return completions.StreamResponsePayload(
+            id=response["id"] if response else None,
```
Comment (harupy) [inline mlflow/gateway/providers/cohere.py:85]: Is `id` only present in the final chunk?

Comment (gabrielfu) [reply mlflow/gateway/providers/cohere.py:85]: yes

File: docs/source/llms/deployments/index.rst [hunk 2]
```diff
@@ -858,7 +858,15 @@ Some providers support streaming responses. Streaming responses are useful when
...
 
-- OpenAI
++------------+---------------------+--------------+
+|  Provider  | Endpoints                          |
++------------+---------------------+--------------+
+|            | llm/v1/completions  | llm/v1/chat  |
++============+=====================+==============+
+| OpenAI     | ✓                   | ✓            |
++------------+---------------------+--------------+
+| Cohere     | ✓                   | ✘            |
```
Comment (harupy) [inline docs/source/llms/deployments/index.rst:868]: Cohere doesn't support chat?

https://docs.cohere.com/reference/chat

File: mlflow/gateway/schemas/completions.py [hunk 3]
```diff
@@ -95,7 +95,7 @@ class StreamChoice(ResponseModel):
 
 
 class StreamResponsePayload(ResponseModel):
-    id: str
+    id: Optional[str]
```
Comment (harupy) [inline mlflow/gateway/schemas/completions.py:98]: Does pydantic automatically set this to None when id is unspecified?

Comment (harupy) [reply mlflow/gateway/schemas/completions.py:98]:
```suggestion
    id: Optional[str] = None
```

Can we explicitly set the default?

File: mlflow/gateway/providers/base.py [hunk 4]
```diff
@@ -73,19 +73,31 @@ def check_for_model_field(payload):
 class ProviderAdapter(ABC):
     """ """
 
-    @abstractclassmethod
+    @classmethod
+    @abstractmethod
```
Comment (harupy) [inline mlflow/gateway/providers/base.py:77]: Nice!

File: docs/source/llms/deployments/index.rst (hunk 2 above)
Comment (gabrielfu) [reply docs/source/llms/deployments/index.rst:868]: We haven't implemented the non-streaming chat endpoint yet. And Cohere chat is in beta. Do you want me to implement both non-streaming & streaming chat?

File: mlflow/gateway/schemas/completions.py (hunk 3 above)
Comment (gabrielfu) [reply mlflow/gateway/schemas/completions.py:98]: Sure

File: docs/source/llms/deployments/index.rst (hunk 2 above)
Comment (harupy) [reply docs/source/llms/deployments/index.rst:868]: I think only the completions API existed when we added Cohere.

> Do you want me to implement both non-streaming & streaming chat?

Yes, if you have bandwidth to do that, but this PR can be merged without them.

File: mlflow/gateway/providers/cohere.py [hunk 5]
```diff
@@ -45,6 +46,63 @@ def model_to_completions(cls, resp, config):
...
+            created=int(time.time()),
+            object="text_completion_chunk",
```
Comment (harupy) [inline mlflow/gateway/providers/cohere.py:87]:
```suggestion
```

Can we remove this line because `object` has the default value?

Comment (gabrielfu) [reply mlflow/gateway/providers/cohere.py]: Sure

File: docs/source/llms/deployments/index.rst (hunk 2 above)
Comment (gabrielfu) [reply docs/source/llms/deployments/index.rst:868]: Ok, let's merge this without them. I'll create a new PR for chat

**Review by harupy — APPROVED** [review]
LGTM!

**Pull request merged**